packages/ejson/lib/src/decoding.dart

```diff
@@ -0,0 +1,272 @@
+// Copyright 2023 MongoDB, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import 'dart:typed_data';
+import 'dart:convert';
+
+import 'package:collection/collection.dart';
+import 'package:ejson_annotation/ejson_annotation.dart';
+import 'package:objectid/objectid.dart';
+import 'package:sane_uuid/uuid.dart';
+import 'package:type_plus/type_plus.dart';
+
+import 'types.dart';
+
+/// Predefined decoders for common types
+const commonDecoders = {
+  dynamic: _decodeAny,
+  Null: _decodeNull,
+  Object: _decodeAny,
+  Iterable: _decodeArray,
+  List: _decodeArray,
+  bool: _decodeBool,
+  DateTime: _decodeDate,
+  Defined: _decodeDefined,
+  Key: _decodeKey,
+  Map: _decodeDocument,
+  double: _decodeDouble,
+  num: _decodeNum,
+  int: _decodeInt,
+  ObjectId: _decodeObjectId,
+  String: _decodeString,
+  Symbol: _decodeSymbol,
+  Uint8List: _decodeBinary,
+  Uuid: _decodeUuid,
+  Undefined: _decodeUndefined,
+  UndefinedOr: _decodeUndefinedOr,
+};
+
+/// Custom decoders for specific types. Use `register` to add a custom decoder.
+final customDecoders = <Type, Function>{};
```

Comment:
Do we want to expose this?

Reply:
There are a handful of members which seem kind of internal in this file and I'm not sure if we should hide them - WDYT?

Reply:
I don't see much harm in keeping them visible, but I'll make `commonDecoders` and `decoders` private, and in `ejson.dart` I'll hide `commonDecoders` and `commonEncoders` from the export, so that the only interface is the `register` function.